src/mlpack/methods/decision_tree/all_categorical_split.hpp

```diff
@@ -34,58 +36,59 @@ class AllCategoricalSplit
   /**
    * Check if we can split a node.  If we can split a node in a way that
    * improves on 'bestGain', then we return the improved gain.  Otherwise we
-   * return the value 'bestGain'.  If a split is made, then classProbabilities
-   * and aux may be modified.  For this particular split type, aux will be empty
-   * and classProbabilities will hold one element---the number of children.
+   * return the value 'bestGain'.  If a split is made, then splitInfo and
+   * aux may be modified.  For this particular split type, aux will be empty
+   * and splitInfo will store the number of children of the node.
    *
    * @param bestGain Best gain seen so far (we'll only split if we find gain
    *      better than this).
    * @param data The dimension of data points to check for a split in.
    * @param numCategories Number of categories in the categorical data.
    * @param labels Labels for each point.
+   * @param begin Start index of labels.
    * @param numClasses Number of classes in the dataset.
    * @param weights Weights associated with labels.
    * @param minimumLeafSize Minimum number of points in a leaf node for
    *      splitting.
-   * @param classProbabilities Class probabilities vector, which may be filled
-   *      with split information a successful split.
+   * @param splitInfo Stores split information on a successful split.
    * @param minimumGainSplit Minimum  gain split.
    * @param aux Auxiliary split information, which may be modified on a
    *      successful split.
    */
-  template<bool UseWeights, typename VecType, typename WeightVecType>
+  template<bool UseWeights, typename VecType, typename ElemType, typename WeightVecType>
   static double SplitIfBetter(
       const double bestGain,
       const VecType& data,
       const size_t numCategories,
-      const arma::Row<size_t>& labels,
+      const arma::Row<ElemType>& labels,
+      const size_t begin,
       const size_t numClasses,
       const WeightVecType& weights,
       const size_t minimumLeafSize,
       const double minimumGainSplit,
-      arma::vec& classProbabilities,
+      double& splitInfo,
```

Comment:
I had used an `arma::vec` here before in case we have more general splitting strategies.  For instance, consider a N-ary numeric split... we'll need an `arma::vec` of length `N - 1` to store each of the split points.

Reply:
Ahh, I see. I thought about it and I think, in that case, we can make it a template. So, in future, if we try to add support for N-ary split, then we can make another function that will store the relevant split information into this variable based on its data type. Does this seem reasonable?

Reply:
I suppose it could work, but it seems like it might be a bit overkill.  You would have to extract the information of what you need to return as a typedef from the split type, and then create one of those and pass it... I dunno, personally I think it's fine just as an `arma::vec`.

Reply:
Actually, If we keep it `arma::vec`, then we will have to write another overload for regression because in that case, we are storing `double splitInfo` for each node in place of `arma::vec classProbabilities`.

So, in my opinion, It would be worth it to add it to make it template. This will help us in two ways - 
1. It will reduce code duplication.
2. It will not affect memory usage or performance either.

I hope I am able to express my point to you. Please let me know your thoughts.

Reply:
Actually, I just now realised, that if in future, we decide to implement N-ary categorical split, then for that, we might need to change the whole bulk of the `SplitIfBetter` and we would also require to pass more information about the split points to the function signature. So, I think that will demand a separate overload of itself. So, maybe we don't have to go the extra mile to tackle that case for now?

Reply:
I'm not sure I understand---if you leave it as an `arma::vec`, you can just use a vector of length 1 to store `double splitInfo`.  This is kind of what happens now... you can think of that `arma::vec` just as some auxiliary storage that can be used for whatever is required.  Maybe we could choose a better name or provide better documentation, but in any case, I guess I am not fully understanding what the issue with leaving it as an `arma::vec` is.

Reply:
`sizeof(arma::vec) = 192` and `sizeof(double) = 8`. So, for each node, we will be wasting 184 bytes. So, for a decently big ensemble of decent size trees, this will make a huge difference. I recall that you once told me that we are trying to reduce the memory of the decision tree as much as possible to decrease the cache miss.

Reply:
Yeah, you are right about that.  But here is the problem that we have... leaves in a decision tree classifier must hold a vector of class probabilities.  This vector could have arbitrary length, as a user's data can have any number of classes.  So we unfortunately can't add the number of classes as a template parameter, and then use a `double[NumClasses]` to store the probabilities.  Thus I chose to use an `arma::vec`.  It's possible we could use something a little bit lighter there, but almost no matter what, we'll end up storing the class probabilities in some allocated memory somewhere else.  (That's what the `arma::vec` will do internally too, if the size of the vector is large enough.)

The other constraint we have is that every leaf node and non-leaf node must be of the same type... so, even if we don't have a leaf, we must still store the member that the leaf uses for the class probabilities.  Therefore, since we already have this `arma::vec classProbabilities` member, I reused it to store the split value for a non-leaf node that was split with `BestBinaryNumericSplit`.  So, at the very least, I don't add an *additional* member that takes up more space.

I think it is an important observation that `sizeof(arma::vec)` is so big, and I do think that perhaps we could improve that, but regardless of what we do, we'll need some kind of member for leaf nodes, so it makes sense to try and reuse that for a non-leaf node.  Maybe it makes sense to approach the `sizeof(arma::vec)` issue separately, in a different PR or something, but, up to you.  Honestly the next best thing I can think is just a `double*` that we manually allocate and deallocate.  It'll be a little bit more overhead, but it'll at least be a lot smaller (should be 8 bytes for the pointer, then whatever we end up allocating will be stored somewhere else).

Reply:
I see. Thanks for the clarification. So, at this moment I think I can make two separate overloads of `AllCategoricalSplit::SplitIfBetter` for classification and regression (the same way I did in `BestBinaryNumericSplit::SplitIfBetter`). This will at least save space for regression, and we can worry about classification some other day.

Reply:
Yes, I guess you are thinking to use just a `double` for regression?  That sounds fine to me, but, I think, for some regression split types that need more than a single `double`, we'll have to use `AuxiliarySplitInfo`.  (The split types you are implementing, I think, don't need extra information.  But, suppose you wanted to have a binary split type for categoricals, that split any categorical feature into two groups of types.  In that case, a single `double` would not be sufficient---you'd need some kind of mapping from each categorical value to indicate whether that category should go to the left or right child.)

Reply:
Up to you too, but maybe we should open a separate issue about `sizeof(classProbabilities)` being so large for classification decision trees?  We don't need to address it immediately, of course, but it might be good to just write down somewhere that it is an issue that we eventually need to handle. :)